File: leetcode/p00205/p00205.py
class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(1)
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        mapping = {}
        seen = {}

        for i in range(len(s)):
            if s[i] in mapping and mapping[s[i]] != t[i]:
                return False
            elif s[i] not in mapping and t[i] in seen:
                return False
            else:
                mapping[s[i]] = t[i]
                seen[t[i]] = True
        
        return True

if __name__ == '__main__':
    s = Solution()

    assert s.isIsomorphic("egg", "add") == True
    assert s.isIsomorphic("foo", "bar") == False
    assert s.isIsomorphic("paper", "title") == True
    assert s.isIsomorphic("abcdefghijklmnopqrstuvwxyzva", "abcdefghijklmnopqrstuvwxyzck") == False
    assert s.isIsomorphic("badc", "baba") == False

File: leetcode/p00205/p00205.go
package p00205

// Time Complexity: O(n)
// Space Complexity: O(1)
func isIsomorphic(s string, t string) bool {
	sPattern, tPattern := map[byte]int{}, map[byte]int{}

	for i := range s {
		if sPattern[s[i]] != tPattern[t[i]] {
			// 兩邊上一次遇到的 index 不一樣, 表示順序不對
			return false
		} else {
			// 將各自的字母紀錄相同的 index
			sPattern[s[i]] = i + 1
			tPattern[t[i]] = i + 1
		}
	}
	return true
}
